course5/homework12/comparers.h

```diff
@@ -0,0 +1,83 @@
+#pragma once
+#include <string>
+#include <cstring>
+#include <cstdio>
+#include <stdexcept>
+#include <boost/tokenizer.hpp>
+#include <memory>
+
+namespace homework12{
+
+class AComparer
+{
+public:
+    bool virtual operator() (const std::string& s1, const std::string& s2) const = 0;
+};
+
+std::unique_ptr<AComparer> CreateComparer(bool numeric, bool keyed, unsigned int column, char separator);
+
+class BasicComparer : public AComparer
+{
+public:
+    bool operator() (const std::string& s1, const std::string& s2) const override
+    {
+        return std::strcmp(s1.c_str(), s2.c_str()) < 0;
+    }
+};
+
+class NumericComparer : public AComparer
+{
+public:
+    bool operator() (const std::string& s1, const std::string& s2) const override
+    {
+        try
+        {
+            auto d1 = std::stod(s1);
+            auto d2 = std::stod(s2);
+            return d1 < d2;
+        }
+        catch(std::invalid_argument&)
+        {
+            return std::strcmp(s1.c_str(), s2.c_str()) < 0;
+        }
+    }
+};
+template <typename TComparer>
+struct ColumnComparer: public AComparer
+{
+public:
+    ColumnComparer(int colNumber, char separator)
+                   :colNumber(colNumber), separator(separator), comparer(TComparer())
+    {
+        if (this->colNumber <= 0)
```

Comment:
Why don't validate params when params are obtained? Then you could get away with a simple ```assert``` here.